docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+("FRUIT_COLOR=yellow", "-f banana.txt", "")
+("FRUIT_COLOR=red", "-f cherry.txt", "--remove-pit")
+```
+
+The JobSpec can be extended to hold a list of parameter tuples (which
+are more easily expressed as a list of lists of individual parameters).
+For example:
+
+```
+apiVersion: extensions/v1beta1
+kind: Job
+...
+spec:
+  completions: 3
+  ...
+  template:
+    ...
+  perCompletionArgs:
+    container: 0
+      -
+        - "-f apple.txt"
+        - "-f banana.txt"
+        - "-f cherry.txt"
+      -
+        - "--remove-seeds"
+        - ""
+        - "--remove-pit"
+  perCompletionEnvVars:
+    - name: "FRUIT_COLOR"
+      - "green"
+      - "yellow"
+      - "red"
+```
+
+However, just providing custom env vars, and not arguments, is sufficient
+for many use cases: parameter can be put into env vars, and then
+substituted on the command line.
+
+#### Comparison
+
+The multiple substitution approach:
+
+- keeps the *per completion parameters* in the JobSpec.
+- Drawback: makes the job spec large for job with thousands of completions. (But for very large jobs, the work-queue style or another type of controller, such as  map-reduce or spark, may be a better fit.)
+
+The index-only approach:
+
+- requires that the user keep the *per completion parameters* in a separate storage, such as a configData or networked storage.
+- Drawback: while in separate storage, they could be mutatated, which would have unexpected effects
+- Drawback: Logic for using index to lookup paramters needs to be in the Pod.
+- Drawback: CLIs and UIs are limited to using the "index" as the identity of a pod
+  from a job.  They cannot easily say, for example `repeated failures on the pod processing banana.txt`.
+
+Which approach to take is not yet decided.  We will describe how to implement both
+cases.
+
+#### PetSets
+
+The [PetSets proposal](https://github.com/kubernetes/kubernetes/pull/18016/files?short_path=61f4179#diff-61f41798f4bced6e42e45731c1494cee) has the same problem, and expores
+several other templating options.  It includes concrete examples, and is worth reading.
+
+The *Index substitution-only* option corresponds roughly to PetSet Proposal 1b.
+The `perCompletionArgs` approach is similar to PetSet Proposal 1e, but more restrictive and thus less verbose.
+
+It would be easier for users if Indexed Job and PetSet are similar where possible.
+However, PetSet differs in several key respects:
+
+- PetSet is for ones to tens of instances.  Indexed job should work with tens of
+  thousands of instances.
+- When you have few instances, you may want to given them pet names.  When you have many
+  instances, you that many instances, integer indexes make more sense.
+- When you have thousands of instances, storing the work-list in the JobSpec
+  is verbose.  For PetSet, this is less of a problem.
+- PetSets (apparently) need to differ in more fields than indexed Jobs.
+
+This differs from PetSet in that PetSet uses names and not indexes.
```

Comment:
Although PetSet *might* use indexes.  We at a minimum need to store a value on the pod that the endpoints controller can read to make the shard id - we *could* have a shared "identity" value that could be interpreted as an index.  However, I'm not clear whether it's required yet.